cpp/src/cider/type/plan/BinaryExpr.cpp

```diff
@@ -27,6 +27,61 @@
 namespace Analyzer {
 using namespace cider::jitlib;
 
+int32_t getBinOpScaledFactor(const SQLOps ops,
+                             const int32_t self_scale,
+                             const int32_t other_scale,
+                             const int32_t res_scale,
+                             const bool first_operand) {
+  switch (ops) {
+    case kPLUS:
+    case kMINUS:
+      return res_scale - self_scale;
+    case kMULTIPLY:
+      return 0;
+    case kDIVIDE:
+    case kMODULO: {
+      if (first_operand)
+        return other_scale + res_scale - self_scale;
+      else
+        return 0;
+    }
+    case kEQ:
+    case kBW_EQ:
+    case kNE:
+    case kLT:
+    case kGT:
+    case kLE:
+    case kGE:
+    case kBW_NE:
+      return std::max(self_scale, other_scale) - self_scale;
+    default:
+      return 0;
+  }
+}
+
+JITValuePointer getRescaledValue(CodegenContext& context,
+                                 FixSizeJITExprValue& expr_val,
+                                 const int32_t scaled_factor) {
+  JITFunction& func = *context.getJITFunction();
+  if (scaled_factor == 0)
+    return expr_val.getValue();
+  // only do upscale to keep precision
+  CHECK_GT(scaled_factor, 0);
+  auto val_tag = expr_val.getValue()->getValueTypeTag();
+  JITValuePointer new_val = func.createVariable(val_tag, "rescaled_val");
+  func.createIfBuilder()
```

Comment:
What about always doing the rescale to avoid if...else... branch here?

Reply:
this is only for upscale overflow check(ignore/skip null value check), and arithmetic with overflow check(WIP fix bugs) will cover in next pr.